components/gitpod-protocol/src/protocol.ts

```diff
@@ -968,9 +977,47 @@ export namespace AdditionalContentContext {
     }
 }
 
-export interface CommitContext extends WorkspaceContext, Commit {
+export interface CommitContext extends WorkspaceContext, GitCheckoutInfo {
     /** @deprecated Moved to .repository.cloneUrl, left here for backwards-compatibility for old workspace contextes in the DB */
     cloneUrl?: string
+
+    /**
+     * The clone and checkout information for additional repositories in case of multi-repo projects.
+     */
+    additionalRepositoryCheckoutInfo?: GitCheckoutInfo[];
+}
+
+export namespace CommitContext {
+    /**
+     * Creates an identifier pointing to the commits of the CommitContext and all sub-repo commit infos.
+     * The identifier is max 255 chars long.
+     * @param commitContext
+     * @returns identifier for set of commits
+     */
+    export function getCommitSHAs(commitContext: CommitContext): string {
+        if (!commitContext.additionalRepositoryCheckoutInfo || commitContext.additionalRepositoryCheckoutInfo.length === 0) {
+            return commitContext.revision;
+        }
+        // it's a multi-repo project, we need to store all commit shas and that needs to fit into 255 chars. Let's only store at most first seven.
+        const commits = [commitContext.revision, ... commitContext.additionalRepositoryCheckoutInfo.map(i => i.revision)];
+        const maxLength = Math.min(7, 255 / commits.length);
```

Comment:
By limiting our SHA length to 7 characters or less, we accept the following risks of collisions (sources: [font-v](https://github.com/source-foundry/font-v/issues/2#issuecomment-327189455), [solid](https://www.getsolid.io/blog/birthday-paradox-coding-solid/), [Wikipedia](https://en.wikipedia.org/wiki/Birthday_problem#Reverse_problem)):

```ts
function collisionRisk (probability, digits) {
  const d = Math.pow(16, digits);
  return Math.sqrt(2 * d * Math.log(1 / (1 - probability)));
}
```

| Probability of collision happening | Number of prebuilds needed with length 7 | with length 6 | with length 5 | with length 4 | with length 3 |
| --- | --- | --- | --- | --- | --- |
| 1% | 2323 | 581 | 146 | 37 | 10 |
| 10% | 7521 | 1881 | 471 | 118 | 30 |
| 50% | 19291 | 4823 | 1206 | 302 | 76 |
| 90% | 35160 | 8790 | 2198 | 550 | 138 |
| 99.99% | 70320 | 17580 | 4395 | 1099 | 275 |

For reference:
- The Gitpod repo has 15355 prebuilds
- The GitLab repo has 79099 prebuilds

If a collision happens, we very subtly break incremental prebuilds, by considering a random old prebuild to be an ancestor even though it's not (e.g. a random old prebuild from a work-in-progress Pull Request suddenly becomes the new base 💣).

Reply:
That should actually have been `Math.*max*(7, ...)` in which case collision probability would be the same as now. But I'll change this to store the full commit in a new field, because I think the parse/unparse contract to squeeze multiple commits into 255 chars is prone to missinterpretation and therefore errors.